daffodil-test/src/test/java/org/goodudfs/example/StringFunctions/StringFunctionsProvider.java

```diff
@@ -0,0 +1,25 @@
+package org.goodudfs.example.StringFunctions;
+
+import org.apache.daffodil.udf.*;
+
+public class StringFunctionsProvider extends UDFunctionProvider {
+	public StringFunctionsProvider() {
+		super.setFunctionClasses( new Class<?>[] { Replace.class, Compare.class } );
+	}
+
+	public Object lookupFunctionClass(String namespace, String name) {
+		Object functionClass = null;
+
+		String nn = String.join("_", namespace, name);
+
+		switch (nn) {
+		case "com.ext.UDFunction.StringFunctions_replace":
```

Comment:
Namespace string doesn't match. 
It would be really nice if we had a way to get these strings off of static methods of the Replace class, or Compare class respectively, rather than having these string constants.

Reply:
Yeah, something like that seems like a good practice. Note that we can't enforce it because in Java you can't have static interfaces and require the user to implement static functions. But as a practice that would be a good idea.

Does it make sense to require a namespace/name static member via reflection, like we require an evaluate function to exist? That would sort of force good conventions. But that duplicates the FunctionClassInfo annotation. Does it make sense to replace the FunctionClassInfo with satic members?

Reply:
Looking at things more thoroughly, it seems a repetetive to need to call setFunctionClasses and provide a lookupFunctionClass. That's two different ways for the user to provide the list of Classes they support. Seems easier for a user to just haave the setFunctionClasses thing to provide the list of classes and then we can query them to get the name/namespace and figure out which one we need to intantiate.

Reply:
@mbeckerle  The namespace string doesn't match to demonstrate that it doesn't need to match the java namespace.The thought process was to give the user the ability to provide a namespace like "http://ext.stringfunctions.com" in the xsd, when the namespace of the udf was something like org.apache.daffodil.udfunctions.javafunctions.Stringfunctions for example. If we don't want to provide this functionality, the lookup is the same for us, we'll call lookupFunctionClass with whatever namespace/name is provided in the schema. It'd be up to the provider to implement the lookup how they wish.

@stevedlawrence So the advantage of having lookupFunctionClass is that it provides a object, which means it can also provide state to the object it initializes. Whereas if we are responsible for initializing the class, we would also be responsible for initializing it with whatever state it might need. Initializing objects can also be a bit messy from reflection, so I thought it best to leave that in the hands of the provider class. It might need a name change as it's literrally supposed to be returning an object, not just a class, like getFunctionClasses would.

I'm not against the static namespace/name methods, and if we choose to provide a UDF abstract class that they have to extend, we can sort of force them to have it, but the checks we'll have to do to make sure it actually returns valid values are not so different for what we do for annotations. So I'm not sure which is better either.

Reply:
Ah, state is a good reason to have the the lookup function. I'm not against a rename if it makes things more clear.

I guess the problem with a UDF abstract class with name/namespace, now that I think about it, is that we would need to create an instance of the UDF to be able to get the information, which we don't want to do. So it must be eiher an annotation of static members. I guess they are basically the same reflection so I'm not sure it matters that much. Maybe the advantage of annotations is those are compile time checked to ensure the right attributes are provided (I assume?). Whereas we can't enforce that static members exist at compile time.

Reply:
So If we go the static members route, we can use reflection to get the fields like we do for the annotations. Then do the same checks we do for the annotated fields. So if you mean schema compile time, we can totally check that it exists.

Reply:
I meant at UDF compile time. If we go the annotation route, I assume that Java will fail to compile if the annotation doesn't have the name/namespace proeprties? If we go the static member route, then we can't tell until we try to compile the schema. Not sure if that's a big win though.

Reply:
This is true, as long as the default value for the annotation properties isn't set (which they are now), and they apply the annotation without filling out the elements. There is no way to enforce the annotations are applied (at all or correctly) at UDF compile time. But if they leave it out, we'd catch it during our checks. So as you've stated annotations provide a wee bit more advantage in this respect.